api: support iproto feature discovery

DifferentialOrange opened the pull request:
api: support iproto feature discovery

Since version 2.10.0 Tarantool supports feature discovery [1]. Client
can send client protocol version and supported features and receive
server protocol version and supported features information to tune
its behavior.

After this patch, the request will be sent on `dial`, before
authentication is performed. Connector stores server info in connection
internals. User can also set option RequiredProtocolVersion and
RequiredProtocolFeatures to fast fail on connect if server does
not provide some expected feature, similar to net.box opts [2]. It is
not clear how connector should behave in case if client doesn't support
a protocol feature or protocol version, see [3]. For now we decided not
to check requirements on client side.

Feature check iterates over lists to check if feature is
enabled. It seems that iterating over a small list is way faster than
building a map, see [4]. Benchmark tests show that this check is rather
fast (0.5 ns for both client and server check on HP ProBook 440 G5) so
it is not necessary to cache it in any way.

Traces of IPROTO_FEATURE_GRACEFUL_SHUTDOWN flag and protocol version 4
could be found in Tarantool source code but they were removed in the
following commits before the release and treated like they never
existed. We also ignore them here too. See [5] for more info. In latest
master commit new feature with code 4 and protocol version 4 were
introduced [6].

1. https://github.com/tarantool/tarantool/issues/6253
2. https://www.tarantool.io/en/doc/latest/reference/reference_lua/net_box/#lua-function.net_box.new
3. https://github.com/tarantool/tarantool/issues/7953
4. https://stackoverflow.com/a/52710077/11646599
5. https://github.com/tarantool/tarantool-python/issues/262
6. https://github.com/tarantool/tarantool/commit/948e5cdce18b081a8f7b03ebd43e34a029b7aefe

Closes #120

oleg-jukovec commented:
> After this patch, the request will be send on `Connect`.

1. What about reconnect?
2. A connection is not necessarily in the connected state after Connect(), see https://github.com/tarantool/go-tarantool/issues/136

> (for now client does not support any features from the list)

I guess it supports IPROTO_FEATURE_STREAMS and IPROTO_FEATURE_TRANSACTIONS at now.

> In this case, protocol version would be ProtocolVersionUnsupported (-1) and no features would be enabled.

We are changing the type here from uint to int. Could we use 0 instead of -1?

> For now it doesn't seems like exposing protocol version and enabled features is useful for a client so private variables are used to store this info. Getters added in export_test for tests.

Why then add this code (feature check) at all? t looks useless.
What is the Tarantool client doing now, how is this information used in it?

oleg-jukovec commented on connection.go:
```diff
@@ -35,6 +35,9 @@ const (
 type ConnLogKind int
 
+var clientProtocolVersion int64 = 3
+var clientFeatures []uint64 = []uint64{}
```
Go has array constants:
https://stackoverflow.com/questions/13137463/declare-a-constant-array

oleg-jukovec commented on connection.go:
```diff
@@ -35,6 +35,9 @@ const (
 type ConnEventKind int
 type ConnLogKind int
 
+var clientProtocolVersion int64 = 3
```
```suggestion
const clientProtocolVersion = 3
```

DifferentialOrange commented:
> Could we use 0 instead of -1?

Well, protocol version is 0 for Tarantool between "version when IPROTO_ID was introduced" and "somewhere between 2.10.0-beta1 and 2.10.0-beta2 when first feature flag was added", so "unknown version" and "version 0" defines different version of Tarantool. On the other hand, it's not like we would treat `-1` and `0` protocol version different in out code.

DifferentialOrange commented:
> I guess it supports IPROTO_FEATURE_STREAMS and IPROTO_FEATURE_TRANSACTIONS at now.

It's an interesting point. I have thought that features won't work until you enable them with feature flag. It is the case for `IPROTO_FEATURE_ERROR_EXTENSION`: until you send the flag, you will receive MP_ERROR objects as plain error Lua tables. But it seems that transactions and streams will work disregarding feature flag.

oleg-jukovec commented on connection.go:
```diff
@@ -146,6 +149,13 @@ type Connection struct {
 
 	lastStreamId uint64
+
+	// protocolVersion is IProto max protocol version supported both by
+	// client and server. Equal to ProtocolVersionUnsupported is server
```
> max protocol version supported both by client and server

What is the idea? How can this be used in practice?
I find it more useful to know the version of the server.

DifferentialOrange replied:
There is no known usage for protocol version yet.

> I find it more useful to know the version of the server.

Well, in case server has newer version than client the only thing we could do is to understand that there are some features we're not know about yet.

oleg-jukovec replied:
> Well, in case server has newer version than client the only thing we could do is to understand that there are some features we do'nt know about yet.

It looks better than nothing.

DifferentialOrange replied:
> > Well, protocol version is 0 for Tarantool between "version when IPROTO_ID was introduced" and "somewhere between 2.10.0-beta1 and 2.10.0-beta2 when first feature flag was added", so "unknown version" and "version 0" defines different version of Tarantool. On the other hand, it's not like we would treat `-1` and `0` protocol version different in out code.
> 
> No, it is not true, see:
> 
> [tarantool/tarantool@0be1faf1fe](https://github.com/tarantool/tarantool/commit/0be1faf1fe)
> 
> It was introduced with `1`.

Sorry for the misinformation then. I'll use `0` as `ProtocolVersionUnsupported`.

DifferentialOrange replied:
Reworked the connector to store server provided info and a handle to check if a feature (or a protocol version) is supported by current connection (server + client).

DifferentialOrange replied:
Close this branch since the discussion continues in https://github.com/tarantool/go-tarantool/pull/226#discussion_r1021682159

DifferentialOrange commented:
> Why then add this code (feature check) at all? t looks useless.
What is the Tarantool client doing now, how is this information used in it?

For example, it enables full MP_ERROR support, see https://github.com/tarantool/go-tarantool/pull/226#issuecomment-1312417403

DifferentialOrange commented:
> the type here from uint to int

Btw, what type should be used to store values like this? `uint8`, `uint16`, `uint32`, `uint64`, `uint`? Some `int` value? Is there any general practice here?

oleg-jukovec commented:
> Well, protocol version is 0 for Tarantool between "version when IPROTO_ID was introduced" and "somewhere between 2.10.0-beta1 and 2.10.0-beta2 when first feature flag was added", so "unknown version" and "version 0" defines different version of Tarantool. On the other hand, it's not like we would treat `-1` and `0` protocol version different in out code.

No, it is not true, see:

https://github.com/tarantool/tarantool/commit/0be1faf1fe

It was introduced with `1`.

oleg-jukovec commented:
> > the type here from uint to int
> 
> Btw, what type should be used to store values like this? `uint8`, `uint16`, `uint32`, `uint64`, `uint`? Some `int` value? Is there any general practice here?

We can hide the implementation details here:
```Go
type ProtocolVersion uint64
```

oleg-jukovec commented:
> > Why then add this code (feature check) at all? t looks useless.
> > What is the Tarantool client doing now, how is this information used in it?
> 
> For example, it enables full MP_ERROR support, see [#226 (comment)](https://github.com/tarantool/go-tarantool/pull/226#issuecomment-1312417403)

Are you sure about this? It looks weird becase MP_ERROR has been introduced before IPROTO_ID.

oleg-jukovec commented on const.go:
```diff
@@ -69,10 +72,17 @@ const (
 	RLimitDrop = 1
 	RLimitWait = 2
 
+	FeatureStreams        = uint64(0)
+	FeatureTransactions   = uint64(1)
+	FeatureErrorExtension = uint64(2)
+	FeatureWatchers       = uint64(3)
```
Seems like a good idea to make a separate enumeration type:

https://github.com/tarantool/go-tarantool/blob/48cf0c75b3869e54fc47881f0a1323612c3694fe/stream.go#L9-L23

+

In general, you could extract some of the pull request code in a separate file (`features.go` ?).

DifferentialOrange replied:
Reworked the code

Totktonada commented:
One of ideas behind the feature discovery functionality is to allow a user to require a feature/a protocol version at connection, so the connection will fail if the server doesn't support it. A kind of fail early approach.

DifferentialOrange commented:
> It looks unexpectedly becase MP_ERROR has been introduced before IPROTO_ID

It seems there were other handles like `msgpack.cfg.encode_error_as_ext` to control `MP_ERROR` support. (But this approach was rather complicated and I failed to create a connection supporting MP_ERROR for pre-2.10.0 Tarantool.)

oleg-jukovec commented on connection.go:
```diff
@@ -391,6 +393,13 @@ func Connect(addr string, opts Opts) (conn *Connection, err error) {
 		}
 	}
 
+	if err = conn.loadProtocolInfo(); err != nil {
+		conn.mutex.Lock()
+		defer conn.mutex.Unlock()
+		conn.closeConnection(err, true)
+		return nil, err
+	}
```
It will be a better place for it:

https://github.com/tarantool/go-tarantool/blob/0081afbcec293ca5044de86164714764a0052915/connection.go#L513

DifferentialOrange replied:
Yeah, that's a nice one. Moved

DifferentialOrange commented:
> 1. What about reconnect?

Added call to `reconnect`, reworked code in `Connect` to be similar to schema load one. I decided not to put feature load code to a single place in `createConnection` since it seems out of place.


> 2\. A connection is not necessarily in the connected state after Connect(), see

To be honest, I don't know if I should do anything with this.

oleg-jukovec commented on connection.go:
```diff
@@ -1163,3 +1176,54 @@ func (conn *Connection) NewStream() (*Stream, error) {
+	return nil
+}
+
+// IsProtocolVersionSupported checks if expected protocol version
+// is supported by both server and client.
+// Since 1.10.0
+func (conn *Connection) IsProtocolVersionSupported(version ProtocolVersion) bool {
+	return (version <= ClientProtocolVersion) &&
+		(version <= conn.serverProtocolInfo.version)
+}
+
+// IsProtocolFeatureSupported checks if expected protocol feature
+// is supported by both server and client.
+// Since 1.10.0
+func (conn *Connection) IsProtocolFeatureSupported(feature ProtocolFeature) bool {
+	return isFeatureSupported(feature, ClientProtocolFeatures) &&
+		isFeatureSupported(feature, conn.serverProtocolInfo.features)
+}
```
This interface hides potentially useful information:

1. Supported protocol version by a server.
2. Supported features by a server.

For example, now it is impossible to check is the server supports MP_ERROR or watchers.

DifferentialOrange replied:
How this info could be useful to a user?

DifferentialOrange replied:
Well, feature flags exist to display if some codes are known (or unknown) to IProto. I'm not sure they are suitable to check Tarantool features disregarding using any client.

I see the reason to write a code like this:
```
if (server_support_watchers and client_support_watchers) {
    -- code with watchers
} else {
    -- code without watchers
}
```

oleg-jukovec replied:
I would still not try to guess the method a use case because we may be wrong.

We have an information about server, let's clearly share it.

DifferentialOrange replied:
Well, ok, that's fair

oleg-jukovec commented on tarantool_test.go:
```diff
@@ -2830,6 +2830,50 @@ func TestStream_DoWithClosedConn(t *testing.T) {
+func TestConnectionProtocolInfo(t *testing.T) {
+	conn := test_helpers.ConnectWithValidation(t, server, opts)
+
+	isProtocolInfoUnsupported, err := test_helpers.IsTarantoolVersionLess(2, 10, 0)
```
It will be better to use conn.IsProtocolVersionSupported() or something like it.

DifferentialOrange commented:
One more point: I've tried to add fast fail conditions in code, but the initial attempt didn't worked out well. The idea was to "return an error if user tries to start using a stream/open a transaction if connection does not support it". But the only way to work with stream in Go code is to allocate `NewStream` which is a simple object allocation without any network requests. And the only way to work with streams is to send a transaction requests through `Do` method, so to check a transaction feature support I must place a check into stream `Do` depending on request encoded type. Both felt really weird to me so for now I've just added an example about how user can ensure that everything is supported.

oleg-jukovec commented:
> The idea was to "return an error if user tries to start using a stream/open a transaction if connection does not support it".

I have not noticed such limitations in the Tarantool implementation. For my taste, you don't have to do this.

But do not to connect to a server without some features looks potentially useful:
https://github.com/tarantool/go-tarantool/pull/226#issuecomment-1312781708

But it may not be possible to implement this in the current reconnection implementation.

oleg-jukovec commented:
> > 1. What about reconnect?
> 
> Added call to `reconnect`, reworked code in `Connect` to be similar to schema load one. I decided not to put feature load code to a single place in `createConnection` since it seems out of place.

1. Schema load is not required.
2. It's a bit more logical to check the protocol and its version first, before any request (authorization). Especially if we implement such restrictions: https://github.com/tarantool/go-tarantool/pull/226#issuecomment-1312781708
 
> > 2. A connection is not necessarily in the connected state after Connect(), see
> 
> To be honest, I don't know if I should do anything with this.

Of course, because otherwise we will have errors in this case:

https://github.com/tarantool/go-tarantool/issues/136

DifferentialOrange commented:
> > To be honest, I don't know if I should do anything with this.
> 
> Of course, because otherwise we will have errors in this case:
> 
> #136

As far as I understand, placing the code to `dial()` solves possible issues since this one is called on both connect/reconnect and we either have a case when connection is not ready or we connection is ready and we had exchanged protocol info.

oleg-jukovec commented on CHANGELOG.md:
```diff
@@ -10,6 +10,8 @@ Versioning](http://semver.org/spec/v2.0.0.html) except to the first release.
 
 ### Added
 
+- Support IProto feature discovery (#120).
```
https://www.tarantool.io/en/doc/latest/dev_guide/internals/box_protocol/
```suggestion
- Support iproto feature discovery (#120).
```

DifferentialOrange replied:
Fixed

oleg-jukovec commented on connection.go:
```diff
@@ -502,6 +505,13 @@ func (conn *Connection) dial() (err error) {
 	conn.Greeting.Version = bytes.NewBuffer(greeting[:64]).String()
 	conn.Greeting.auth = bytes.NewBuffer(greeting[64:108]).String()
 
+	// IPROTO_ID requests can be processed without authentication.
+	// https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/requests/#iproto-id
+	if err = conn.loadProtocolInfo(w, r); err != nil {
+		connection.Close()
+		return err
+	}
+
```
What about https://github.com/tarantool/go-tarantool/pull/226#issuecomment-1312781708 ?

DifferentialOrange replied:
Implemented in new version

oleg-jukovec commented on connection.go:
```diff
@@ -581,43 +591,63 @@ func pack(h *smallWBuf, enc *encoder, reqid uint32,
 
 	if err != nil {
-		return errors.New("auth: pack error " + err.Error())
+		return errors.New(reqName + ": pack error " + err.Error())
```
Here you can not use the argument reqName, but wrap the error in a caller code https://go.dev/blog/go1.13-errors , see `Wrapping errors with %w`

The result will be something like that:

```Go
if err := writeRequest(w, req); err != nil {
    return fmt.Errorf("failed to send iproto id: %w", err)
}
```

DifferentialOrange replied:
Thank you, it's much better. Fixed.

oleg-jukovec commented on connection.go:
```diff
@@ -581,43 +591,63 @@ func pack(h *smallWBuf, enc *encoder, reqid uint32,
+	return conn.writeRequest(w, req, "iproto id")
+}
+
+func (conn *Connection) readResponse(r io.Reader, reqName string) (resp Response, err error) {
```
Same as for `writeRequest` (error wrapping).

DifferentialOrange replied:
Fixed

oleg-jukovec commented on connection.go:
```diff
@@ -1163,3 +1193,71 @@ func (conn *Connection) NewStream() (*Stream, error) {
+	return conn.serverProtocolVersion
+}
+
+// ClientProtocolVersion returns protocol version supported by
+// Go connection client.
+// Since 1.10.0
+func (conn *Connection) ClientProtocolVersion() ProtocolVersion {
+	return ClientProtocolVersion
+}
```
Why not just use a constant ClientProtocolVersion? It looks like duplicate functionality.

oleg-jukovec replied:
Let's discuss it here:

https://github.com/tarantool/go-tarantool/pull/226#discussion_r1022628684

oleg-jukovec commented on connection.go:
```diff
@@ -1163,3 +1193,71 @@ func (conn *Connection) NewStream() (*Stream, error) {
+	return isFeatureSupported(feature, conn.serverFeatures)
+}
+
+// DoesClientSupportFeature checks if expected feature
+// is supported by Go connection client.
+// Since 1.10.0
+func (conn *Connection) DoesClientSupportFeature(feature Feature) bool {
+	return isFeatureSupported(feature, ClientFeatures)
+}
```
Why not just check values from ClientFeatures? It looks like duplicate functionality.

DifferentialOrange replied:
It would be a bit inconvenient to have asymmetrical getters for client and server info. [2]
Also it is possible that we would enable/disable some features for a connection, like MP_ERROR encoding mode.

oleg-jukovec replied:
> Also it is possible that we would enable/disable some features for a connection, like MP_ERROR encoding mode, so it became connection-dependent.

Then we need to give this opportunity, but hide ClientProtocolVersion and ClientFeatures from the public API.

DifferentialOrange replied:
Well, to be honest, I don't get how I should design this one.

There should be a set of features
- provided by user
- required by user

and I don't see any way to mix them. Moreover, I don't see any way to actually allow user to provide a set of features. For example, they may not provide (intentionally or non-intentionally) `TransactionFeature` flag, but client actually supports this feature and we even provide the API for it. Either we start to disallow to use feature considering feature flag set or we allow user to misinform the Tarantool server. (For now, it is hard to say if this may have any consequences.)

Same goes about protocol version. To be honest, it is rather hard to design a solution for a such strange thing.

For now it seems that
- 3 flags doesn't do anything
- 1 flag changes the behavior when both client and server provides it

and, to be honest, I've planned to enable/disable error extension flag the following way:
- enable if user exports `_ = github.com/tarantool/go-tarantool/error_extension`
- do not enable otherwise.

DifferentialOrange replied:
So it seems that for now we decided to not provide a user handles for feature setup. Constants are hidden now

oleg-jukovec commented on connection.go:
```diff
@@ -1163,3 +1193,71 @@ func (conn *Connection) NewStream() (*Stream, error) {
+// DoesServerSupportFeature checks if expected feature
+// is supported by connected Tarantool server.
+// Since 1.10.0
+func (conn *Connection) DoesServerSupportFeature(feature Feature) bool {
```
`Is` feels more idiomatic.

```suggestion
func (conn *Connection) IsServerSupportFeature(feature Feature) bool {
```

DifferentialOrange replied:
It is more idiomatic, but it is a mistake in terms of English language grammar.
![image](https://user-images.githubusercontent.com/20455996/201900029-de56219a-7f4f-4f54-b17a-f434f8dc2146.png)

oleg-jukovec replied:
> but it is a mistake in terms of English language grammar

I don't think it's important in this context (a call that returns a boolean value).

DifferentialOrange replied:
Well, it's not relevant anymore

oleg-jukovec commented on connection.go:
```diff
@@ -1163,3 +1193,71 @@ func (conn *Connection) NewStream() (*Stream, error) {
+// is supported by connected Tarantool server.
+// Since 1.10.0
+func (conn *Connection) DoesServerSupportFeature(feature Feature) bool {
+	return isFeatureSupported(feature, conn.serverFeatures)
```
Data race is another potential problem of such checks: reconnect to a server with a lower Tarantool version. The rules from this idea look more strict and are less prone to such situations:

https://github.com/tarantool/go-tarantool/pull/226#issuecomment-1312781708

DifferentialOrange replied:
Implemented in new version

oleg-jukovec commented on connection_pool/connection_pool.go:
```diff
@@ -1028,3 +1028,47 @@ func newErrorFuture(err error) *tarantool.Future {
 	return fut
 }
+
+// ServerProtocolVersion returns protocol version supported by
+// Tarantool server for a connection selected by userMode from connPool.
+// Since 1.10.0
+func (connPool *ConnectionPool) ServerProtocolVersion(userMode Mode) (tarantool.ProtocolVersion, error) {
+	conn, err := connPool.getNextConnection(userMode)
+	if err != nil {
+		return tarantool.ProtocolVersionUnsupported, err
+	}
+	return conn.ServerProtocolVersion(), nil
+}
```
1. We have multi submodule too.
2. I don't understand how to use it. Well, I found out that one of the connections supports something, so what of it?

It looks more useful in this context:
https://github.com/tarantool/go-tarantool/pull/226#issuecomment-1312781708

DifferentialOrange replied:
Implemented in new version

oleg-jukovec commented on protocol.go:
```diff
@@ -0,0 +1,44 @@
+
+const ProtocolVersionUnsupported ProtocolVersion = 0
+
+const (
+	// Streams support.
+	FeatureStreams Feature = 0
+	// Interactive transactions support.
+	FeatureTransactions Feature = 1
+	// Support of MP_ERROR object over MessagePack.
+	FeatureErrorExtension Feature = 2
+	// Support of watchers.
+	FeatureWatchers Feature = 3
+)
```
```suggestion
const (
	// Streams support.
	StreamsFeature Feature = 0
	// Interactive transactions support.
	TransactionsFeature Feature = 1
	// Support of MP_ERROR object over MessagePack.
	ErrorExtensionFeature Feature = 2
	// Support of watchers.
	WatchersFeature Feature = 3
)
```

Let's unify it here:

https://github.com/tarantool/go-tarantool/blob/master/stream.go#L9-L23
https://github.com/tarantool/go-tarantool/blob/master/datetime/adjust.go#L5-L11
https://github.com/tarantool/go-tarantool/blob/master/connection_pool/state.go#L8-L14

oleg-jukovec commented on request.go:
```diff
@@ -1106,3 +1106,60 @@ func (req *ExecuteRequest) Context(ctx context.Context) *ExecuteRequest {
+
+// protocolInfoRequest informs the server about supported protocol
+// version and features.
+type protocolInfoRequest struct {
```
1. I don't see anything wrong with making this request public.
2. We need tests for the request, see request_test.go

DifferentialOrange replied:
Added tests and publicity. Should we also test sending this request in runtume? (As a plain request that doesn't update any internals)

oleg-jukovec replied:
Sure.

DifferentialOrange replied:
Added request and request ctx test

oleg-jukovec commented on request.go:
```diff
@@ -1106,3 +1106,60 @@ func (req *ExecuteRequest) Context(ctx context.Context) *ExecuteRequest {
+
+// protocolInfoRequest informs the server about supported protocol
+// version and features.
+type protocolInfoRequest struct {
```
```suggestion
type idRequest struct {
```

oleg-jukovec reviewed (changes requested):
I think we need to implement this feature for Connection/ConnectionPool/ConnectionMulti.

https://github.com/tarantool/go-tarantool/pull/226#issuecomment-1312781708

+

Calls for manual check of server features for ConnectionPool and ConnectionMulti seems useless for me.

DifferentialOrange commented:
> Calls for manual check of client/server features for ConnectionPool and ConnectionMulti seems useless for me.

So in the end we will not provide any getters for server info?

oleg-jukovec commented:
> So in the end we will not provide any getters for server info?

It can be useful for `Connection`, but it is necessary to aware a user that the result may be outdated in disconnected state.

In the case of `ConnectionPool` and `ConnectionMulti` it will be just calls without real benefits.

So let's left getters for server info for `Connection`, but remove them from `ConnectionPool` and `ConnectionMulti` + implement the feature https://github.com/tarantool/go-tarantool/pull/226#issuecomment-1312781708 for `Connection`, `ConnectionPool` and `ConnectionMulti`

oleg-jukovec commented on connection.go:
```diff
@@ -351,6 +364,17 @@ func Connect(addr string, opts Opts) (conn *Connection, err error) {
+	conn.clientProtocolVersion = ClientProtocolVersion
+	conn.clientFeatures = ClientFeatures
+
+	if err = checkProtocolVersion(opts.RequiredProtocolVersion, conn.clientProtocolVersion); err != nil {
+		return nil, fmt.Errorf("client: %w", err)
+	}
+
+	if err = checkFeatures(opts.RequiredFeatures, conn.clientFeatures); err != nil {
+		return nil, fmt.Errorf("client: %w", err)
+	}
+
```
It is also better to use a copy of the slice opts.RequiredFeatures.

DifferentialOrange replied:
Well, to be honest, it seems more like an excessive copy since this is an internal method and we know what it may do, but let it be. Changed

oleg-jukovec replied:
```Go
someFeaturesList = []Feature{Feature1, Feature2}
opts.RequiredFeatures = someFeatureList
conn := tarantool.Connect(conn, opts)
opts.RequiredFeatures[0] = Feature666 // It will change 
```

DifferentialOrange replied:
Added test

oleg-jukovec commented on protocol.go:
```diff
@@ -0,0 +1,80 @@
+	WatchersFeature Feature = 3
+)
+
+var featureName = map[Feature]string{
+	StreamsFeature:        "StreamsFeature",
+	TransactionsFeature:   "TransactionsFeature",
+	ErrorExtensionFeature: "ErrorExtensionFeature",
+	WatchersFeature:       "WatchersFeature",
+}
```
It will be more universal to add String() method for Feature type.

```Go
func (f Feature) String() string {
    switch f {
    StreamsFeature:
        return "StreamsFeature"
    ...
    default:
        return fmt.Sprintf("unknown feature %d", f)
    }
}
```

DifferentialOrange replied:
Thank you, fixed

oleg-jukovec commented on protocol.go:
```diff
@@ -0,0 +1,80 @@
+	TransactionsFeature,
+}
+
+func unsupportedFeatureError(feature Feature) error {
+	name, ok := featureName[feature]
+	if !ok {
+		return fmt.Errorf("feature with code %d is not supported", feature)
+	}
+
+	// "Feature" already is a part of the name.
+	return fmt.Errorf("%s is not supported", name)
+}
+
+func checkProtocolVersion(expected ProtocolVersion, actual ProtocolVersion) error {
+	if expected > actual {
+		return fmt.Errorf("protocol version %d is not supported", expected)
+	}
+
+	return nil
+}
+
+func checkFeatures(expectedList []Feature, actualList []Feature) error {
+	var found bool
+	// It seems that iterating over a small list is way faster
+	// than building a map: https://stackoverflow.com/a/52710077/11646599
+	for _, expected := range expectedList {
+		found = false
+		for _, actual := range actualList {
+			if expected == actual {
+				found = true
+			}
+		}
+		if !found {
+			return unsupportedFeatureError(expected)
+		}
+	}
+
+	return nil
+}
```
1. It looks like implementation details related to `Connection`
2. You have types, you can add methods to them. To make it more general, public and tested.

DifferentialOrange replied:
1. Moved to connection.
2. I think I cannot add method to a slice, isn't it?

oleg-jukovec commented on protocol_test.go:
```diff
@@ -0,0 +1,65 @@
+package tarantool
```
It is advisable not to test the internal implementation while possible. In this case, we can only test the public API (perhaps by adding useful methods to it).

DifferentialOrange replied:
Well, I don't quite get how unit testing for internal implementation is a bad approach. Isn't it a basic design feature?

oleg-jukovec replied:
It is a very debatable question:

1. https://medium.com/@thrawn01/why-you-should-never-test-private-methods-f822358e010
2. https://roadfiresoftware.com/2018/05/can-i-unit-test-private-methods/

To put it very briefly: it complicates the modification of the implementation.

Yes, it can be a different approach, but in this repository used a simple rules:

1. We are testing only the public interface.
2. If you really want to test integration with something (SSL, msgpack).
3. If you don't want to copy-paste or duplicate too much and it is very easy (better to avoid).

You could see such reasons in `export_test.go` files.

DifferentialOrange replied:
Well, since these unit tests now not cover anything exceptional, removed

oleg-jukovec commented on request.go:
```diff
@@ -1106,3 +1130,35 @@ func (req *ExecuteRequest) Context(ctx context.Context) *ExecuteRequest {
 	return req
 }
+
+// IdRequest informs the server about supported protocol
+// version and features.
+type IdRequest struct {
+	baseRequest
+	idInfo
+}
+
+// newIdRequest returns a new IdRequest.
+func newIdRequest(version ProtocolVersion, features []Feature) *IdRequest {
+	req := new(IdRequest)
+	req.requestCode = IdRequestCode
+	req.version = version
+	req.features = features
+	return req
+}
+
+// Body fills an encoder with the id request body.
+func (req *IdRequest) Body(res SchemaResolver, enc *encoder) error {
+	return fillId(enc, req.version, req.features)
+}
+
+// Context sets a passed context to the request.
+//
+// Pay attention that when using context with request objects,
+// the timeout option for Connection does not affect the lifetime
+// of the request. For those purposes use context.WithTimeout() as
+// the root context.
+func (req *IdRequest) Context(ctx context.Context) *IdRequest {
+	req.ctx = ctx
+	return req
+}
```
Looks like it can be transferred to protocol.go

DifferentialOrange replied:
But it is the code related to requests. And as far as I know, all requests-related code is in `requests.go`.

oleg-jukovec replied:
It seems to me a good idea not to make a dump for requests in the `requests.go`. Specialized requests look better in separate files (see `stream.go`, `prepared.go`).

oleg-jukovec commented on response.go:
```diff
@@ -190,6 +196,24 @@ func (resp *Response) decodeBody() (err error) {
 				if bindCount, err = d.DecodeUint64(); err != nil {
 					return err
 				}
+			case KeyVersion:
+				isIdResponse = true
+				if err = d.Decode(&serverIdInfo.version); err != nil {
+					return err
+				}
+			case KeyFeatures:
+				isIdResponse = true
+				if larr, err = d.DecodeArrayLen(); err != nil {
+					return err
+				}
+
+				serverIdInfo.features = make([]Feature, larr)
+				for i := 0; i < larr; i++ {
+					if err = d.Decode(&feature); err != nil {
+						return err
+					}
+					serverIdInfo.features[i] = feature
+				}
```
https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/requests/#box-protocol-id

KeyVersion + KeyFeatures are required. Should there be an error if something is missing?

DifferentialOrange replied:
Added an error

oleg-jukovec commented on tarantool_test.go:
```diff
@@ -2830,6 +2831,115 @@ func TestStream_DoWithClosedConn(t *testing.T) {
+		conn.ClientFeatures(),
+		[]Feature{StreamsFeature, TransactionsFeature})
+
+	if isLess210 {
+		require.Equal(t, conn.ServerProtocolVersion(), NoProtocolVersion)
+		require.ElementsMatch(t, conn.ServerFeatures(), []Feature{})
+	} else {
+		require.GreaterOrEqual(t,
+			conn.ServerProtocolVersion(),
+			tarantool210ProtocolVersion)
+
+		require.Subset(t,
+			conn.ServerFeatures(),
+			[]Feature{
+				StreamsFeature,
+				TransactionsFeature,
+				ErrorExtensionFeature,
+				WatchersFeature})
+	}
```
It would be easier to split the test into two.

DifferentialOrange replied:
Changed

oleg-jukovec commented on request.go:
```diff
@@ -1106,3 +1130,35 @@ func (req *ExecuteRequest) Context(ctx context.Context) *ExecuteRequest {
+
+// IdRequest informs the server about supported protocol
+// version and features.
+type IdRequest struct {
+	baseRequest
+	idInfo
+}
+
+// newIdRequest returns a new IdRequest.
+func newIdRequest(version ProtocolVersion, features []Feature) *IdRequest {
```
Request type is public, but a user can not to create an object in the right way (`NewIdRequest`).

DifferentialOrange replied:
Changed

oleg-jukovec commented on request.go:
```diff
@@ -1106,3 +1130,35 @@ func (req *ExecuteRequest) Context(ctx context.Context) *ExecuteRequest {
+// the timeout option for Connection does not affect the lifetime
+// of the request. For those purposes use context.WithTimeout() as
+// the root context.
+func (req *IdRequest) Context(ctx context.Context) *IdRequest {
+	req.ctx = ctx
+	return req
+}
```
The code is not covered by any test.

DifferentialOrange replied:
Added request test

DifferentialOrange commented on connection.go:
```diff
@@ -1163,3 +1244,64 @@ func (conn *Connection) NewStream() (*Stream, error) {
+	return conn.clientProtocolVersion
+}
+
+// ServerFeatures returns features supported by connected Tarantool server.
+// Since 1.10.0
+func (conn *Connection) ServerFeatures() []Feature {
+	return conn.serverFeatures
+}
+
+// ClientFeatures returns features supported by Go connection client.
+// Since 1.10.0
+func (conn *Connection) ClientFeatures() []Feature {
+	return conn.clientFeatures
+}
```
> It's better to return a copy of slices

To be honest, I have thought that I already return a copy.

DifferentialOrange replied:
Added

DifferentialOrange commented on connection.go:
```diff
@@ -581,43 +622,83 @@ func pack(h *smallWBuf, enc *encoder, reqid uint32,
 
 	if err != nil {
-		return errors.New("auth: pack error " + err.Error())
+		return errors.New("pack error " + err.Error())
```
Comparing to previous changes, this one may actually change error message. But it's up to you as a maintainer to decide.

DifferentialOrange replied:
Changed

DifferentialOrange commented on connection.go:
```diff
@@ -1163,3 +1244,64 @@ func (conn *Connection) NewStream() (*Stream, error) {
+	return conn.clientProtocolVersion
+}
+
+// ServerFeatures returns features supported by connected Tarantool server.
```
Added

oleg-jukovec commented on example_test.go:
```diff
@@ -26,6 +26,14 @@ func example_connect() *tarantool.Connection {
 	return conn
 }
 
+func example_connect_with_opts(opts tarantool.Opts) *tarantool.Connection {
+	conn, err := tarantool.Connect(server, opts)
+	if err != nil {
+		panic("Connection is not established: " + err.Error())
+	}
+	return conn
+}
```
It will be more obvious from the outside to add arguments to example_connect():

```Go
func example_connect(server string, opts tarantool.Opts) {
}
```

At now it is not very clear what and where to change at first glance. Up to you.

DifferentialOrange replied:
Changed

oleg-jukovec commented on connection.go:
```diff
@@ -502,6 +522,25 @@ func (conn *Connection) dial() (err error) {
+		return fmt.Errorf("server: %w", err)
+	}
+
+	requiredFeatures := make([]Feature, len(opts.RequiredFeatures))
+	copy(requiredFeatures, opts.RequiredFeatures)
```
dial() may be called much later of Connect() (on the reconnect case), see the problem:

https://github.com/tarantool/go-tarantool/pull/226#discussion_r1025243600

It is better to save a copy of the slice at the beginning of Connect() call.

DifferentialOrange replied:
Added opts copy

oleg-jukovec commented on connection.go:
```diff
@@ -351,6 +361,16 @@ func Connect(addr string, opts Opts) (conn *Connection, err error) {
+
+	requiredFeatures := make([]Feature, len(opts.RequiredFeatures))
+	copy(requiredFeatures, opts.RequiredFeatures)
+	if err = checkFeatures(requiredFeatures, conn.ClientFeatures()); err != nil {
+		return nil, fmt.Errorf("client: %w", err)
+	}
```
Let's remove the client features check. Still, the idea is to require something from the server.

DifferentialOrange replied:
Well, it seems like Tarantool check on both sides (even though it's not really clear through source code):

```
Tarantool 2.11.0-entrypoint-700-gc4b8debee
type 'help' for interactive help
tarantool> box.cfg{listen=3301}
```

```
Tarantool 2.10.0-beta2-283-gb56e012bc
type 'help' for interactive help
tarantool> net_box = require('net.box')
---
...

tarantool> net_box.connect('localhost:3301', {required_protocol_features = {'pagination'}})
---
- peer_protocol_version: 4
  peer_uuid: 7735b4a0-fb8c-4530-afe0-0853888f5342
  opts:
    required_protocol_features:
    - pagination
  host: localhost
  peer_protocol_features:
    transactions: true
    watchers: true
    error_extension: true
    streams: true
  protocol: Binary
  state: error
  error: 'Missing required protocol features: pagination'
  peer_version_id: 133888
  port: '3301'
...
```

(While we were reviewing, new feature had been dropped.)

oleg-jukovec replied:
I think `net.box` just can't find `pagination` in an internal mapping. I don't see the point in this in our case: we just limit a user in this case.

DifferentialOrange replied:
> I don't see the point in the check in our case: we just limit a user to know some info about server.

I don't like the case when user requires a feature, gets a success and then finds out that we can't really use it. But it seems that, in case of both net.box and a connector, they would fail when they try to use some handles that are not exist (even better in case of Go since it would fail in compile-time). Ok, let's remove it and file a ticket to tarantool/tarantool about inconsistency.

oleg-jukovec reviewed (commented):
There are a couple of non-critical moments left.

oleg-jukovec commented on connection.go:
```diff
@@ -581,43 +620,83 @@ func pack(h *smallWBuf, enc *encoder, reqid uint32,
 
 	if err != nil {
-		return errors.New("auth: pack error " + err.Error())
+		return fmt.Errorf("pack error %w", err)
```
To avoid surprises, let's do it right as in the examples:
https://go.dev/blog/go1.13-errors

```suggestion
		return fmt.Errorf("pack error: %w", err)
```

and above.

In practice there is no difference:
https://play.golang.com/p/CHwfiMpL5DH
but it is more obvious from a client side.

DifferentialOrange replied:
Changed

Totktonada commented:
For one of discussions above:

> So it seems that for now we decided to not provide a user handles for feature setup. Constants are hidden now

But it contradicts with https://github.com/tarantool/go-tarantool/pull/226#issuecomment-1312781708:

> One of ideas behind the feature discovery functionality is to allow a user to require a feature/a protocol version at connection, so the connection will fail if the server doesn't support it. A kind of fail early approach.

What is the point to don't allow to fail early?

AFAIU you have problems with implementing extended errors in a backward compatible way. Is it the point? Can you explain it a bit for me?

oleg-jukovec commented:
> > One of ideas behind the feature discovery functionality is to allow a user to require a feature/a protocol version at connection, so the connection will fail if the server doesn't support it. A kind of fail early approach.

It is implemented: a user can require a feature/a protocol version for a server.

> > So it seems that for now we decided to not provide a user handles for feature setup.

It is not impelemented: we did not allow the user to set up protocol version (IPROTO_VERSION) and features (IPROTO_FEATURES) for a connection manually:
https://www.tarantool.io/en/doc/latest/dev_guide/internals/iproto/requests/#box-protocol-id
The connector does it himself and sends a supported version/all supported features.

In general, the behavior is similar to `net.box` implementation except a one detail:

The `net.box` implementation checks features (but not version) for a client too, not only for the server. We decided to check only the server because the features of the client are known.

oleg-jukovec commented on connection.go:
```diff
@@ -269,6 +272,14 @@ type Opts struct {
 	Transport string
 	// SslOpts is used only if the Transport == 'ssl' is set.
 	Ssl SslOpts
+	// Minimal protocol version that should be supported by
+	// Go connection client and Tarantool server. By default
+	// it is equal to 0 (no restrictions)
+	RequiredProtocolVersion ProtocolVersion
+	// List of protocol features that should be supported by
+	// Go connection client and Tarantool server. By default
+	// it is a nil array (no restrictions)
+	RequiredProtocolFeatures []ProtocolFeature
```
Could we replace it with?

RequiredServerInfo PrtocolInfo

DifferentialOrange replied:
Yeah, it easily can be

DifferentialOrange replied:
Replaced

oleg-jukovec commented on connection.go:
```diff
@@ -1163,3 +1240,107 @@ func (conn *Connection) NewStream() (*Stream, error) {
+	return conn.serverProtocolInfo.Version
+}
+
+// ClientProtocolVersion returns protocol version supported by
+// Go connection client.
+// Since 1.10.0
+func (conn *Connection) ClientProtocolVersion() ProtocolVersion {
+	return сlientProtocolVersion
+}
```
I think it would be a good idea to add an example test:

func ExampleConnection_ClientProtocolVersion() {
    conn := example_conn(server, opts)
    fmt.Println("Connector client version:", conn.ClientProtocolVersion())
    // Output:
    // 3
}

This will help the user understand what the connector really support.

DifferentialOrange replied:
Added

oleg-jukovec commented on connection.go:
```diff
@@ -1163,3 +1240,107 @@ func (conn *Connection) NewStream() (*Stream, error) {
+	return res
+}
+
+// ClientFeatures returns protocol features supported by Go connection client.
+// Since 1.10.0
+func (conn *Connection) ClientProtocolFeatures() []ProtocolFeature {
+	res := make([]ProtocolFeature, len(сlientProtocolFeatures))
+	copy(res, сlientProtocolFeatures)
+	return res
+}
```
I think it would be a good idea to add an example test [2]. The test will help the user understand what the connector really support.

oleg-jukovec commented on protocol.go:
```diff
@@ -0,0 +1,112 @@
+	Features []ProtocolFeature
+}
+
+const (
+	// Streams support.
+	StreamsFeature ProtocolFeature = 0
+	// Interactive transactions support.
+	TransactionsFeature ProtocolFeature = 1
+	// Support of MP_ERROR objects over MessagePack.
+	ErrorExtensionFeature ProtocolFeature = 2
+	// Support of watchers.
+	WatchersFeature ProtocolFeature = 3
+	// Support of pagination.
+	PaginationFeature ProtocolFeature = 4
+)
```
```suggestion
const (
	// Streams support (supported).
	StreamsFeature ProtocolFeature = 0
	// Interactive transactions support (supported).
	TransactionsFeature ProtocolFeature = 1
	// Support of MP_ERROR objects over MessagePack (unsupported).
	ErrorExtensionFeature ProtocolFeature = 2
	// Support of watchers (unsupported).
	WatchersFeature ProtocolFeature = 3
	// Support of pagination (unsupported).
	PaginationFeature ProtocolFeature = 4
)
```

DifferentialOrange replied:
Added more detailed description

oleg-jukovec commented on protocol.go:
```diff
@@ -0,0 +1,112 @@
+	"fmt"
+)
+
+type ProtocolVersion uint64
+type ProtocolFeature uint64
+
+type ProtocolInfo struct {
+	Version  ProtocolVersion
+	Features []ProtocolFeature
+}
```
Let's add a description to types/fields because this is a public API.

DifferentialOrange replied:
Added

oleg-jukovec reviewed (approved):
Thank you for the patch! Do not wait anything from me after the fixes above.

oleg-jukovec commented on connection.go:
```diff
@@ -1163,3 +1240,107 @@ func (conn *Connection) NewStream() (*Stream, error) {
+// connected Tarantool server. Beware that values might be outdated
+// if connection is in a disconnected state.
+// Since 1.10.0
+func (conn *Connection) ServerProtocolVersion() ProtocolVersion {
+	return conn.serverProtocolInfo.Version
+}
```
It may be appropriate to merge ServerProtocolVersion() + ServerProtocolFeatures() = ServerProtocolInfo() and ClientProtocolVersion() + ClientProtocolFeatures() = ClientProtocolInfo()

DifferentialOrange replied:
Fixed

AnaNek reviewed (approved):
LGTM. Thank you for the patch, see my minor remarks.

oleg-jukovec commented on connection.go:
```diff
@@ -293,6 +300,13 @@ type SslOpts struct {
 	Ciphers string
 }
 
+func (opts Opts) Copy() (optsCopy Opts) {
+	optsCopy = opts
+	optsCopy.RequiredProtocolInfo = opts.RequiredProtocolInfo.Copy()
+
+	return optsCopy
+}
```
```suggestion
func (opts Opts) Copy() Opts {
	optsCopy := opts
	optsCopy.RequiredProtocolInfo = opts.RequiredProtocolInfo.Copy()

	return optsCopy
}
```

DifferentialOrange replied:
Reworked

oleg-jukovec commented on connection.go:
```diff
@@ -581,43 +607,83 @@ func pack(h *smallWBuf, enc *encoder, reqid uint32,
 }
 
-func (conn *Connection) readAuthResponse(r io.Reader) (err error) {
+func (conn *Connection) writeAuthRequest(w *bufio.Writer, scramble []byte) (err error) {
```
```suggestion
func (conn *Connection) writeAuthRequest(w *bufio.Writer, scramble []byte) error {
```

DifferentialOrange replied:
Reworked

oleg-jukovec commented on connection.go:
```diff
@@ -581,43 +607,83 @@ func pack(h *smallWBuf, enc *encoder, reqid uint32,
+	return nil
+}
+
+func (conn *Connection) writeIdRequest(w *bufio.Writer,
+	protocolInfo ProtocolInfo) (err error) {
```
```suggestion
func (conn *Connection) writeIdRequest(w *bufio.Writer,
	protocolInfo ProtocolInfo) error {
```

DifferentialOrange replied:
Reworked

oleg-jukovec commented on connection.go:
```diff
@@ -581,43 +607,83 @@ func pack(h *smallWBuf, enc *encoder, reqid uint32,
+	return nil
+}
+
+func (conn *Connection) readResponse(r io.Reader) (resp Response, err error) {
```
```suggestion
func (conn *Connection) readResponse(r io.Reader)(Response, error) {
```

DifferentialOrange replied:
Reworked

oleg-jukovec commented on connection.go:
```diff
@@ -581,43 +607,83 @@ func pack(h *smallWBuf, enc *encoder, reqid uint32,
+	return resp, nil
+}
+
+func (conn *Connection) readAuthResponse(r io.Reader) (err error) {
```
```suggestion
func (conn *Connection) readAuthResponse(r io.Reader) error {
```

DifferentialOrange replied:
Reworked

oleg-jukovec commented on connection.go:
```diff
@@ -581,43 +607,83 @@ func pack(h *smallWBuf, enc *encoder, reqid uint32,
+	return nil
+}
+
+func (conn *Connection) readIdResponse(r io.Reader) (resp Response, err error) {
```
```suggestion
func (conn *Connection) readIdResponse(r io.Reader) (Response, error) {
```

DifferentialOrange replied:
Reworked

oleg-jukovec commented:
In general, it would be better to avoid naked return values unless there is a need for `defer`.

https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

But this is a controversial topic:

https://github.com/golang/go/issues/21291

DifferentialOrange commented on connection.go:
```diff
@@ -581,43 +607,83 @@ func pack(h *smallWBuf, enc *encoder, reqid uint32,
 }
 
-func (conn *Connection) writeAuthRequest(w *bufio.Writer, scramble []byte) (err error) {
+func (conn *Connection) writeRequest(w *bufio.Writer, req Request) (err error) {
```
Reworked

DifferentialOrange commented on tarantool_test.go:
```diff
@@ -2830,6 +2841,315 @@ func TestStream_DoWithClosedConn(t *testing.T) {
+	err = test_helpers.RestartTarantool(&inst)
+	require.Nilf(t, err, "Failed to restart tarantool")
+
+	connected := test_helpers.WaitUntilReconnected(conn, retries, timeout)
```
Yeah, we definitely need. Thank you for noticing!

DifferentialOrange commented on connection.go:
```diff
@@ -1163,3 +1229,87 @@ func (conn *Connection) NewStream() (*Stream, error) {
+}
+
+// identify sends info about client protocol, receives info
+// about server protocol in response and store it in the connection.
```
Fixed

DifferentialOrange commented on connection.go:
```diff
@@ -1163,3 +1229,87 @@ func (conn *Connection) NewStream() (*Stream, error) {
+
+	if len(missingFeatures) == 1 {
+		return fmt.Errorf("protocol feature %s is not supported", missingFeatures[0])
+	} else if len(missingFeatures) > 1 {
```
Reworked

DifferentialOrange commented on example_test.go:
```diff
@@ -407,7 +429,17 @@ func ExampleRollbackRequest() {
+			tarantool.StreamsFeature,
+			tarantool.TransactionsFeature,
+		},
+	}
```
Reworked

oleg-jukovec commented on protocol.go:
```diff
@@ -0,0 +1,131 @@
+	Features []ProtocolFeature
+}
+
+func (info ProtocolInfo) Copy() ProtocolInfo {
+	infoCopy := info
+
+	if info.Features != nil {
+		infoCopy.Features = make([]ProtocolFeature, len(info.Features))
+		copy(infoCopy.Features, info.Features)
+	}
+
+	return infoCopy
+}
```
Is there a unit test for the method?

DifferentialOrange replied:
Added

oleg-jukovec commented on connection.go:
```diff
@@ -293,6 +300,13 @@ type SslOpts struct {
 	Ciphers string
 }
 
+func (opts Opts) Copy() Opts {
+	optsCopy := opts
+	optsCopy.RequiredProtocolInfo = opts.RequiredProtocolInfo.Copy()
+
+	return optsCopy
+}
```
Is there a unit test for the method?

DifferentialOrange replied:
Added

oleg-jukovec commented on protocol.go:
```diff
@@ -0,0 +1,131 @@
+	"context"
+	"fmt"
+)
+
+type ProtocolVersion uint64
+type ProtocolFeature uint64
+
+type ProtocolInfo struct {
+	// Version is the supported protocol version.
+	Version ProtocolVersion
+	// Features are supported protocol features.
+	Features []ProtocolFeature
+}
+
+func (info ProtocolInfo) Copy() ProtocolInfo {
```
Types ProtocolVersion, ProtocolFeature, ProtocolInfo and the method Copy() has no description, but it is a public API.

DifferentialOrange replied:
Added

oleg-jukovec commented on connection.go:
```diff
@@ -293,6 +300,13 @@ type SslOpts struct {
 	Ciphers string
 }
 
+func (opts Opts) Copy() Opts {
```
The method Copy() has no description, but it is a public API.

DifferentialOrange replied:
Added

oleg-jukovec commented on protocol.go:
```diff
@@ -0,0 +1,131 @@
+	// (unsupported by connector).
+	PaginationFeature ProtocolFeature = 4
+)
+
+func (ftr ProtocolFeature) String() string {
```
No description.

DifferentialOrange replied:
Added

oleg-jukovec commented on test_helpers/utils.go:
```diff
@@ -66,3 +87,31 @@ func SkipIfStreamsUnsupported(t *testing.T) {
 		t.Skip("Skipping test for Tarantool without streams support")
 	}
 }
+
+func SkipIfIdUnsupported(t *testing.T) {
+	t.Helper()
+
+	// Tarantool supports Id requests since version 2.10.0
+	isLess, err := IsTarantoolVersionLess(2, 10, 0)
+	if err != nil {
+		t.Fatalf("Could not check the Tarantool version")
+	}
+
+	if isLess {
+		t.Skip("Skipping test for Tarantool without id requests support")
+	}
+}
+
+func SkipIfIdSupported(t *testing.T) {
```
There are no description.

DifferentialOrange replied:
Added

oleg-jukovec commented on protocol.go:
```diff
@@ -0,0 +1,131 @@
+	PaginationFeature ProtocolFeature = 4
+)
+
+func (ftr ProtocolFeature) String() string {
```
Is there a unit test for the method?

DifferentialOrange replied:
Added

oleg-jukovec commented on protocol_test.go:
```diff
@@ -0,0 +1,59 @@
+		})
+}
+
+func TestOptsCopyPreservesRequiredProtocolFeatures(t *testing.T) {
```
It will look better in another file: connection_test.go ?

DifferentialOrange replied:
Ok. At first I had put it in the separate file, but then decided not to increase the quantity of files in the root folder. (But it seems that it is ok in Go.)

DifferentialOrange replied:
Moved

oleg-jukovec commented on connection.go:
```diff
@@ -293,6 +300,17 @@ type SslOpts struct {
 	Ciphers string
 }
 
+// Copy returns the copy of an Opts object.
```
I suggest to use `Clone` instead of `Copy`. It may be more obvious.

https://pkg.go.dev/golang.org/x/exp/slices#Clone
https://pkg.go.dev/strings#Clone

DifferentialOrange replied:
Thank you, changed

oleg-jukovec commented on connection.go:
```diff
@@ -293,6 +300,17 @@ type SslOpts struct {
 	Ciphers string
 }
 
+// Clone returns a copy of the Opts object.
+// Beware that Notify channel, Logger and Handle are not copied.
```
It seems confusing:
```suggestion
// Clone returns a copy of the Opts object.
// Beware that Notify channel, Logger and Handle are not copied.
```
because they are copied in fact:
```
opts.Close().Notify == opts.Notify
```

DifferentialOrange replied:
Removed `// Beware that Notify channel, Logger and Handle are not copied.` line

oleg-jukovec reviewed (approved):
I'm ready to merge if you resolve with @AnaNek all conversations.

DifferentialOrange commented:
> I'm ready to merge if you resolve with @AnaNek all conversations.

Conversations are resolved now.

oleg-jukovec commented:
> > I'm ready to merge if you resolve with @AnaNek all conversations.
> 
> Conversations are resolved now.

```
 internal: copy opts on Connect

The copy is not an honest deepcopy because, for example, copying logger
or channel will break the logic.
```
It seems better to rename `Copy()` -> `Close()` in the commit instead of a next one.

DifferentialOrange commented:
> > > I'm ready to merge if you resolve with @AnaNek all conversations.
> > 
> > 
> > Conversations are resolved now.
> 
> ```
>  internal: copy opts on Connect
> 
> The copy is not an honest deepcopy because, for example, copying logger
> or channel will break the logic.
> ```
> 
> It seems better to rename `Copy()` -> `Clone()` in the commit instead of a next one.

Sorry, I missed it. Fixed

oleg-jukovec commented on connection.go:
```diff
@@ -269,6 +272,10 @@ type Opts struct {
 	Ssl SslOpts
+	// RequiredProtocolInfo contains minimal protocol version and
+	// list of protocol features that should be supported by
+	// Tarantool server. By default there are no restrictions
```
```suggestion
	// Tarantool server. By default there are no restrictions.
```

DifferentialOrange replied:
Fixed

oleg-jukovec commented:
Thank you!
